pkg/operator/encryptionprovider/encryptionprovider.go

```diff
@@ -0,0 +1,93 @@
+package encryptionprovider
+
+import (
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/sets"
+	corev1listers "k8s.io/client-go/listers/core/v1"
+
+	"github.com/openshift/cluster-openshift-apiserver-operator/pkg/operator/operatorclient"
+	"github.com/openshift/library-go/pkg/operator/encryption/controllers"
+	"github.com/openshift/library-go/pkg/operator/encryption/encryptionconfig"
+	"github.com/openshift/library-go/pkg/operator/events"
+	operatorv1helpers "github.com/openshift/library-go/pkg/operator/v1helpers"
+)
+
+type encryptionProvider struct {
+	oauthAPIServerTargetNamespace   string
+	oauthEncryptionCfgAnnotationKey string
+
+	currentEncryptedGRs                 []schema.GroupResource
+	defaultEncryptedGRs                 []schema.GroupResource
+	encryptedGRsManagedByExternalServer sets.String
+
+	secretLister corev1listers.SecretNamespaceLister
+
+	eventRecorder events.Recorder
+}
+
+var _ controllers.Provider = &encryptionProvider{}
+
+func NewEncryptionProvider(
+	oauthAPIServerTargetNamespace string,
+	oauthEncryptionCfgAnnotationKey string,
+	defaultEncryptedGRs []schema.GroupResource,
+	encryptedGRsManagedByExternalServer sets.String,
+	kubeInformersForNamespaces operatorv1helpers.KubeInformersForNamespaces,
+	eventRecorder events.Recorder) *encryptionProvider {
+	return &encryptionProvider{
+		oauthAPIServerTargetNamespace:       oauthAPIServerTargetNamespace,
+		oauthEncryptionCfgAnnotationKey:     oauthEncryptionCfgAnnotationKey,
+		defaultEncryptedGRs:                 defaultEncryptedGRs,
+		encryptedGRsManagedByExternalServer: encryptedGRsManagedByExternalServer,
+		secretLister:                        kubeInformersForNamespaces.InformersFor(operatorclient.GlobalMachineSpecifiedConfigNamespace).Core().V1().Secrets().Lister().Secrets(operatorclient.GlobalMachineSpecifiedConfigNamespace),
+		eventRecorder:                       eventRecorder,
+	}
+}
+
+// EncryptedGRs returns resources that need to be encrypted
+func (p *encryptionProvider) EncryptedGRs() []schema.GroupResource {
+	oauthAPIServerEncryptionCfg, err := p.secretLister.Get(fmt.Sprintf("%s-%s", encryptionconfig.EncryptionConfSecretName, p.oauthAPIServerTargetNamespace))
+	if err != nil {
+		return p.defaultEncryptedGRs
+	}
+
+	if _, exist := oauthAPIServerEncryptionCfg.Annotations[p.oauthEncryptionCfgAnnotationKey]; exist {
+		return p.defaultEncryptedGRs
+	}
+
+	newEncryptedGRsToManage := []schema.GroupResource{}
+	for _, gr := range p.defaultEncryptedGRs {
+		if p.encryptedGRsManagedByExternalServer.Has(gr.String()) {
+			continue
+		}
+		newEncryptedGRsToManage = append(newEncryptedGRsToManage, gr)
+	}
+	if changed, newEncryptedGRsToManageSet := haveGRsChanged(p.currentEncryptedGRs, newEncryptedGRsToManage); changed {
+		p.eventRecorder.Eventf("EncryptedGRsChanged", "The new GroupResource list this operator will manage is %v", newEncryptedGRsToManageSet.List())
+		p.currentEncryptedGRs = newEncryptedGRsToManage
```

Comment:
this is wrong spot. Also state like this is an anti-pattern. Side-effects from a getter like this are very unexpected.

Reply:
which getter? the value is read from the cache and won't change once set.

Reply:
the state (`currentEncryptedGRs`) was introduced to emit an event only once it has changed.

Reply:
> this is wrong spot.

why ? we assign to `currentEncryptedGRs` only when it has changed.

Reply:
Wasn't clear with my comment. Emitting events from a getter like this seems to be wrong.

Reply:
wouldn't it be nice to know that the list have changed (debuggability)? 
frequent events might indicate errors.